test/specs/extension_type_test.dart

```diff
@@ -0,0 +1,242 @@
+          ..name = 'bar')),
+      equalsDart(r'''
+        extension type Foo<T,U>(T bar) { }
+      '''),
+    );
+  });
+
+  test('extension type with generics bound', () {
+    expect(
+      ExtensionType((b) => b
+        ..name = 'Foo'
+        ..types.add(TypeReference((b) => b
+          ..symbol = 'T'
+          ..bound = TypeReference((b) => b..symbol = 'num')))
+        ..representationDeclaration = RepresentationDeclaration((b) => b
+          ..declaredRepresentationType = TypeReference((b) => b.symbol = 'T')
+          ..name = 'bar')),
+      equalsDart(r'''
+        extension type Foo<T extends num>(T bar) { }
+      '''),
+    );
+  });
+
+  test('extension type with named primary constructor', () {
+    expect(
+      ExtensionType((b) => b
+        ..name = 'Foo'
+        ..primaryConstructorName = 'named'
+        ..representationDeclaration = RepresentationDeclaration((b) => b
+          ..declaredRepresentationType = TypeReference((b) => b.symbol = 'int')
+          ..name = 'bar')),
+      equalsDart(r'''
+        extension type Foo.named(int bar) { }
+      '''),
+    );
+  });
+
+  test('extension type with metadata on field', () {
+    expect(
+      ExtensionType((b) => b
+        ..name = 'Foo'
+        ..representationDeclaration = RepresentationDeclaration((b) => b
+          ..declaredRepresentationType = TypeReference((b) => b.symbol = 'int')
+          ..name = 'bar'
+          ..docs.add(
+            '/// My favorite representation declaration.',
+          )
+          ..annotations.addAll([
+            refer('deprecated'),
+            refer('Deprecated').call(
+                [literalString('This is an old representation declaration')])
+          ]))),
+      equalsDart(r'''
+        extension type Foo(/// My favorite representation declaration.
+          @deprecated
+          @Deprecated('This is an old representation declaration')
+          int bar) { }
+      '''),
+    );
+  });
+
+  test('extension type with implements', () {
+    expect(
+      ExtensionType((b) => b
+        ..name = 'Foo'
+        ..implements.add(TypeReference((b) => b.symbol = 'num'))
+        ..representationDeclaration = RepresentationDeclaration((b) => b
+          ..declaredRepresentationType = TypeReference((b) => b.symbol = 'int')
+          ..name = 'bar')),
+      equalsDart(r'''
+        extension type Foo(int bar) implements num { }
+      '''),
+    );
+  });
+
+  test('extension type with multiple implements', () {
+    expect(
+      ExtensionType((b) => b
+        ..name = 'Foo'
+        ..implements.addAll([
+          TypeReference((b) => b.symbol = 'num'),
+          TypeReference((b) => b.symbol = 'Object')
+        ])
+        ..representationDeclaration = RepresentationDeclaration((b) => b
+          ..declaredRepresentationType = TypeReference((b) => b.symbol = 'int')
+          ..name = 'bar')),
+      equalsDart(r'''
+        extension type Foo(int bar) implements num,Object { }
```

Comment:
I would expect a space between these, the matcher might just be a bit loose though because I thought you did add that, but might as well update the expectation to show it:

```suggestion
        extension type Foo(int bar) implements num, Object { }
```

Reply:
This was the one change I made before the force push. :)

I changed the separator to not have that space to be consistent with the rest of the emitter and to avoid spaces that aren't needed in the output. Dart format should add the space for us when 2.3.4 is released.

Reply:
Ah ok, SGTM :)